### PR TITLE
Implement multilabel attention in CNN

### PR DESCRIPTION
**Pull request by nsorros**
Description
---

The idea is that each label will have a separate self attention mechanism that will attend to the inputs differently. This sounds like an expensive operation but it is quite cheap as each attention requires a vector of the same dimension as the words for example `400`. Of course we need 30k of those but this not different to the dimensionality of the embedding layer which has a vocabulary size of 30k as well.

See a schematic of multi-label attention from this [paper](https://papers.nips.cc/paper/2019/file/9e6a921fbc428b5638b3986e365d4f21-Paper.pdf) below

Screenshot 2021-07-28 at 12 54 48 PM

Checklist
---

- [ ] Added link to Github issue or Trello card
- [x] Added tests

### CODE REVIEW

**Comment by nsorros** [comment]
This is also ready for review @aCampello

**Comment by nsorros** [comment]
This PR would love some feedback 😬 @aCampello

**Comment by nsorros** [comment]
The last commit did not seem to have triggered the tests 🤔

**Comment by nsorros** [comment]
This waits for a wonderful review @aCampello

**Comment by aCampello** [comment]
Apologies for taking embarrassingly long, will give it a go today.

File: wellcomeml/ml/cnn.py [hunk 1]
```diff
@@ -233,11 +240,17 @@ def residual_attention(x1):
         output_activation = (
             "sigmoid" if nb_outputs == 1 or self.multilabel else "softmax"
         )
+        if self.feature_approach == "multilabel-attention":
+            # each output comes from an attention head
+            nb_outputs = 1
         out = tf.keras.layers.Dense(
             nb_outputs,
```
Comment (aCampello) [inline wellcomeml/ml/cnn.py:247]: I don't understand, if this is multilabel attention, the dense layer will only have one unit?

Comment (aCampello) [reply wellcomeml/ml/cnn.py:247]: Ah, I think I worked out the maths. The layer will be `(batch_size, number_of_labels, 1)`

File: wellcomeml/ml/cnn.py [hunk 2]
```diff
@@ -233,11 +240,17 @@ def residual_attention(x1):
         output_activation = (
             "sigmoid" if nb_outputs == 1 or self.multilabel else "softmax"
         )
+        if self.feature_approach == "multilabel-attention":
+            # each output comes from an attention head
+            nb_outputs = 1
```
Comment (aCampello) [inline wellcomeml/ml/cnn.py:245]: It's a bit weird that you use nb_outputs as it's actual value for the output activation, then you change it then use it again, but that's just style

**Review by aCampello — APPROVED** [review]
I really apologise for embarrassingly dropping the ball here. I was supposed to review this change while you were on holiday, but the business got in the way.


Notice that in the [paper](https://papers.nips.cc/paper/2019/file/9e6a921fbc428b5638b3986e365d4f21-Paper.pdf), it seems that the model is _learning_ the hierarchy of labels.

Now in your case you already have the tree structure - it comes from MeSH! It seems that there should be way of leveraging that instead.

File: wellcomeml/ml/cnn.py (hunk 1 above)
Comment (nsorros) [reply wellcomeml/ml/cnn.py:247]: 👌

File: wellcomeml/ml/cnn.py (hunk 2 above)
Comment (nsorros) [reply wellcomeml/ml/cnn.py:245]: I understand this is confusing, will try to improve terminology in the future

**Comment by nsorros** [comment]
> I really apologise for embarrassingly dropping the ball here. I was supposed to review this change while you were on holiday, but the business got in the way.
> 
> Notice that in the [paper](https://papers.nips.cc/paper/2019/file/9e6a921fbc428b5638b3986e365d4f21-Paper.pdf), it seems that the model is _learning_ the hierarchy of labels.
> 
> Now in your case you already have the tree structure - it comes from MeSH! It seems that there should be way of leveraging that instead.

Indeed they use a tree but not from MeSH. Constructing that tree might be something we will have to look soon but the idea behind it is described here http://cs.uef.fi/sipu/pub/Balanced_kmeans_ssspr14.pdf

**Comment by aCampello** [comment]
Feel free to merge. I think the new numpy causes the pipeline to fail, we'll see.

**Comment by nsorros** [comment]
This is ready to merged, I mean psychologically ready
